unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let (depth, name) = match link.Name {
+        Some(Cow::Borrowed(s)) if s.len() > 2 => (depth, s[2..].to_owned()),
+        Some(Cow::Borrowed(s)) if s.len() == 2 => (depth - 1, String::from("")),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => {
+            if s.len() == 2 {
+                (depth - 1, String::from(""))
+            } else {
+                assert!(s.len() > 2);
+                (depth, s[2..].to_owned())
+            }
+        }
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+impl Walker {
+    /// Returns a new instance of a walker, ready to start from the given Cid.
+    pub fn new(cid: Cid, root_name: String) -> Walker {
+        // 1 == Path::ancestors().count() for an empty path
+        let depth = if root_name.is_empty() { 1 } else { 2 };
+        let next = Some((cid, root_name, depth));
+
+        Walker {
+            current: None,
+            next,
+            pending: Vec::new(),
+        }
+    }
+
+    /// Returns a description of a kind of node Walker is currently looking at, if the
+    /// continue_walk has been called after creating the value.
+    pub fn as_entry(&'_ self) -> Option<Entry<'_>> {
+        self.current.as_ref().map(|c| c.as_entry())
+    }
+
+    /// Returns the next cid to load and pass content of which to pass to `continue_walk`.
+    pub fn pending_links<'a>(&'a self) -> (&'a Cid, impl Iterator<Item = &'a Cid> + 'a) {
+        use InnerKind::*;
+        // rev: because we'll pop any of the pending
+        let cids = self.pending.iter().map(|(cid, ..)| cid).rev();
+
+        match self.current.as_ref().map(|c| &c.kind) {
+            Some(File(_, Some(ref visit), _)) => {
+                let (first, rest) = visit.pending_links();
+                let next = self.next.iter().map(|(cid, _, _)| cid);
+                (first, Either::Left(rest.chain(next.chain(cids))))
+            }
+            _ => {
+                let next = self
+                    .next
+                    .as_ref()
+                    .expect("validated in start and continue_walk we have the next");
+                (&next.0, Either::Right(cids))
+            }
+        }
+    }
+
+    /// Continues the walk.
+    ///
+    /// Returns a descriptor for the next element found as `ContinuedWalk` which includes the means
+    /// to further continue the walk. `bytes` is the raw data of the next block, `cache` is an
+    /// optional cache for data structures which can always be substituted for `&mut None`.
+    pub fn continue_walk<'a>(
+        mut self,
+        bytes: &'a [u8],
+        cache: &mut Option<Cache>,
+    ) -> Result<ContinuedWalk<'a>, Error> {
+        use InnerKind::*;
+
+        if let Some(File(_, visit @ Some(_), _)) = &mut self.current.as_mut().map(|c| &mut c.kind) {
```

Comment:
is the leading `&mut` needed here?

Reply:
Clippy does not complain about this
```suggestion
        if let Some(File(_, visit @ Some(_), _)) = self.current.as_mut().map(|c| &mut c.kind) {
```

Apparently not. Not caught with clippy.